### PR TITLE
Feature/3521 Enable purely iteration based training

### LINKED ISSUE #3521: Enable training purely based on number of iterations instead of epochs
## 🚀 Feature
Enable training purely based on number of iterations instead of epochs

### Motivation

This can be useful for certain training runs. Without this feature, the user must set an unreachably high value for `max_epochs` and set `max_steps` to the desired iteration count. With this setup, the trainer will break from the training loop based on `max_steps` since we'd never reach `max_epochs`. For example, [Detectron2](https://github.com/facebookresearch/detectron2/blob/master/detectron2/engine/train_loop.py#L125-L147) uses iteration-based training in their train loop.

### Pitch
The solution could be pretty simple. We can make `min_epochs` and `max_epochs` as Optional. If all of min/max_epochs and max_steps are unset, use the defaults we have today (min_epochs=1, max_epochs=1000). If only `max_steps` is set, use that (and keep `min/max_epochs` as None). If all are set, stop based on whichever condition is hit first.

Specifically, we'd need to:
- Initialize things correctly here: https://github.com/PyTorchLightning/pytorch-lightning/blob/c64520e658806a87f282c74299b2ea4b7ea493ea/pytorch_lightning/trainer/training_loop.py#L46
- Update the training loop here to not depend on `max_epochs`: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L340-L375 . 

Are there other spots I'm missing? 

### Alternatives

Without touching the Trainer code, a super hacky solution would be setting `max_epochs` to be an outrageously large value and set `max_steps` to my desired iteration count. then we'll break from the train loop according to `max_steps` since we'd meet this condition first instead of `max_epochs`. However this could be better handled

### PR DESCRIPTION
**Pull request by richardqiu**
## What does this PR do?

Fixes #3521 

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
can you set your Black to ignore ' and " with `-S` or `--skip-string-normalization`?

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -59,8 +60,8 @@ def on_trainer_init(self, max_epochs, min_epochs, max_steps, min_steps, num_sani
         self.trainer.train_dataloader = None
         self.automatic_optimization = automatic_optimization
 
-        self.trainer.max_epochs = max_epochs
-        self.trainer.min_epochs = min_epochs
+        self.trainer.max_epochs = 1000 if (max_epochs is None and max_steps is None) else max_epochs
```
Comment (justusschock) [inline pytorch_lightning/trainer/training_loop.py:63]: This seems to be the super hacky solution @ananthsub mentioned in the issue. Can we instead use the more clean one that also accepts None here and simply does not break the loop?

**Review by justusschock — CHANGES_REQUESTED** [review]

**Comment by williamFalcon** [comment]
@richardqiu can you write a completely separate test and try not to modify the existing tests?

A good place to put the test is in trainer/flags/test_min_max_steps_epochs.py

**Review by williamFalcon — CHANGES_REQUESTED** [review]
please add a different test and try not to touch current tests

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (richardqiu) [reply pytorch_lightning/trainer/training_loop.py:63]: Apologies if I'm misunderstanding since I'm not very familiar with the code base--my intention for this was to set `max_epochs` to 1000 if neither `max_epochs` nor `max_steps` were specified. If only `max_steps` is specified, then `max_epochs` will be None and the iterator [here](https://github.com/PyTorchLightning/pytorch-lightning/pull/4086/files#diff-0cd7170e96efa4a55577086f5db2fe4cR475-R476) will handle that appropriately.

**Comment by tchaton** [comment]
Hey @richardqiu,

Would you mind updating your PR ?

Best,
T.C

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/training_loop.py:65]: @richardqiu do you mind adding this as a comment in the code, so folks reading this will see that 1000 is the existing default for `max_epochs`?

File: pytorch_lightning/trainer/__init__.py [hunk 2]
```diff
@@ -876,8 +876,9 @@ def on_train_end(self, trainer, pl_module):
 
 |
 
-Stop training after this number of steps
-Training will stop if max_steps or max_epochs have reached (earliest).
+Stop training after this number of steps.
+If both max_epochs and max_steps are specified, training will stop if either max_steps or max_epochs have been
+reached (earliest).
```
Comment (ananthsub) [inline pytorch_lightning/trainer/__init__.py:881]: the documentation for max_epochs above needs to be updated too

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -475,7 +473,8 @@ def train(self):
 
         try:
             # run all epochs
-            for epoch in range(self.current_epoch, self.max_epochs):
+            epochs = range(self.current_epoch, self.max_epochs) if self.max_epochs else count(self.current_epoch)
+            for epoch in epochs:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:477]: what happens if `self.max_epochs` is `None`, and `self.max_steps` is large enough that we make more than 1 pass through the data?  the `count(self.current_epoch)` would return an iterable over a single value right? does that mean the loop would stop too early?

File: tests/trainer/flags/test_min_max_steps_epochs.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    """
+    model = BoringModel()
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        min_epochs=0,
+        max_steps=3,
+        min_steps=0,
+        weights_summary=None,
+    )
```
Comment (ananthsub) [inline tests/trainer/flags/test_min_max_steps_epochs.py:18]: could you also test with a larger max_steps value? i want to test a case where `trainer.current_epoch` should be > 1 even if `max_epochs` is `None`

File: tests/trainer/flags/test_min_max_steps_epochs.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    model = BoringModel()
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        min_steps=3,
+        max_epochs=2,
+        weights_summary=None,
+    )
+
+    trainer.fit(model)
```
Comment (ananthsub) [inline tests/trainer/flags/test_min_max_steps_epochs.py:41]: - are there other min/max epoch test cases that could be moved into this file?
- we should comprehensively test all possibilities for min/max epochs/steps. i think that's 8 test cases, which you could parameterize

**Review by ananthsub — CHANGES_REQUESTED** [review]

**Comment by justusschock** [comment]
@richardqiu May I ask you to also include #4406 to this, since this comes hand in hand with this PR?

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (richardqiu) [reply pytorch_lightning/trainer/trainer.py:477]: Don't believe so, unless I'm misunderstanding? the `count()` iterator is indefinite/doesn't raise `StopIteration`.

**Comment by richardqiu** [comment]
Sorry about the delay here folks! I've added the extra tests that @ananthsub asked for (might have missed a few cases though? let me know!). 

@justusschock happy to add #4406 to the PR. Is the idea that we could do validation every n batches/steps, where n > (number of batches/training epoch)?

**Comment by richardqiu** [comment]
Sorry about the delay here folks! I've added the extra tests @ananthsub asked for, but I might have missed some--let me know!

And I'd be happy to add #4406 @justusschock. Is the idea just that we should allow validation every `n` training batches/steps, where `n > (number of batches/steps per training epoch)`? If that's the case, what should we do for `n < (number of batches/steps per training epoch)`? Should we use the current behavior (where the batch count resets every batch)? Or should we persist the count across epochs and use `trainer.global_step` for both cases?

**Comment by justusschock** [comment]
@richardqiu I thought the idea was not to have epochs any more. So you would just have the maximum number of steps and every n steps you run validation.

**Comment by Borda** [comment]
@richardqiu @ananthsub @justusschock how is it going here, ready to land? 🐰

**Comment by richardqiu** [comment]
There are two failing tests in `tests/trainer/test_dataloaders.py` involving the changes to `check_val_interval`, which I'm not sure how to resolve. Advice here would be appreciated!

**Comment by justusschock** [comment]
@richardqiu This is also coming from the rebasing when we introduced lightning optimizer in #4658 . Seems like the rebasing skipped the added arguments there.

**Comment by hadim** [comment]
@richardqiu are you still working on this PR?

**Comment by Borda** [comment]
@richardqiu as it is a feature I am changing the target branch to feat 1.2, mind rebase pr and resolve conflicts? Thx your contribution and feel free to ping us back if you need something or need some help... 🐰

**Comment by ananthsub** [comment]
@Borda I'll take this up and re-submit it

**Comment by kaushikb11** [comment]
> @Borda I'll take this up and re-submit it

@ananthsub I believe this PR could resolve this [Issue](https://github.com/PyTorchLightning/pytorch-lightning/issues/4809) as well. Do let me know if I could help here.

**Comment by richardqiu** [comment]
> @Borda I'll take this up and re-submit it

Thanks, I appreciate it! Apologies for the long delays here folks, and thanks for the feedback I've gotten here so far!
